core/src/main/java/life/qbic/identityaccess/application/user/UserRegistrationService.java

```diff
@@ -137,81 +139,60 @@ private RegistrationResponse validateInput(String fullName, String email, char[]
-      var successResponse = new RegistrationResponse();
-      successResponse.setType(Type.SUCCESSFUL);
-      return successResponse;
-    }
-
-    public static RegistrationResponse failureResponse(RuntimeException... exceptions) {
-      if (exceptions == null) {
-        throw new IllegalArgumentException("Null references are not allowed.");
-      }
-      var failureResponse = new RegistrationResponse();
-      failureResponse.setType(Type.FAILED);
-      failureResponse.setExceptions(exceptions);
-      return failureResponse;
-    }
-
-    private RegistrationResponse() {
-      super();
-    }
-
-    private void setType(Type type) {
-      this.type = type;
-    }
-
-    public Type getType() {
-      return type;
-    }
-
-    private void setExceptions(RuntimeException... exceptions) {
-      this.exceptions = Arrays.stream(exceptions).toList();
+  /**
+   * Requests a password reset for a user.
+   *
+   * @param userId the user id of the user for whom the password reset shall be issued
+   * @return application response with success or failure information
+   * @since 1.0.0
+   */
+  public ApplicationResponse requestPasswordReset(String userId) {
+    UserId id;
+    try {
+      id = UserId.from(userId);
+    } catch (IllegalArgumentException e) {
+      return ApplicationResponse.failureResponse(e);
     }
-
-    public boolean hasFailures() {
-      return type == Type.FAILED;
+    // fetch user
+    var optionalUser = userRepository.findById(id);
+    if (optionalUser.isEmpty()) {
+      return ApplicationResponse.failureResponse(new UserNotFoundException());
     }
 
-    public List<RuntimeException> failures() {
-      return exceptions;
-    }
+    // trigger password reset
+    var user = optionalUser.get();
+    DomainEventPublisher.instance().subscribe(new DomainEventSubscriber<PasswordReset>() {
+      @Override
+      public Class<? extends DomainEvent> subscribedToEventType() {
+        return PasswordReset.class;
+      }
 
-    /**
-     * Depending on the response, two type of downstream actions can be passed to the
-     * {@link RegistrationResponse}.
-     * <p>
-     * If the instance contains failures, the downstream failure {@link Consumer} action will be
-     * triggered and a reference to itself passed as argument. Otherwise, the downstream failure
-     * consumer is called.
-     *
-     * @param downstreamSuccess consumer for success responses
-     * @param downstreamFailure consumer for failure responses
-     * @since 1.0.0
-     */
-    public void ifSuccessOrElse(Consumer<RegistrationResponse> downstreamSuccess,
-        Consumer<RegistrationResponse> downstreamFailure) {
-      if (hasFailures()) {
-        downstreamFailure.accept(this);
-      } else {
-        downstreamSuccess.accept(this);
+      @Override
+      public void handleEvent(PasswordReset event) {
+        eventStore.append(event);
+        sendNotification(event, notificationService);
       }
-    }
+    });
```

Comment:
because the application service interacts with the domain and is responsible for the orchestration of downstream process actions after a domain event happened. 
The domain event happens in the same thread as the application service is running, but we still need to broadcast the event, which is not domain concern.

Reply:
Ok I am not sure if I fully understand. My point is, that I think this part can be implemented maybe in the constructor or in another method called in the constructor.

Reply:
I need to go back to Vaughn Vernon's chapter about the concept of implementing Domain Events and come back to you to answer your question

Reply:
So, after some literature research, I am still convinced that this is the good way to do it. We have an transactional boundary that is created when the client calls the application service. The service then operates on the domain model and creates subscribers to Domain Events. In our case this is only one, but potentially there can be more than one subscribers necessary to form eventual consistency in our model.